enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,549 @@
+mutate regular pods.
+
+Kubelet will be changed so that when the feature is enabled, it
+advertises a new extended resource of "management cores", representing
+all of the CPU capacity of the host (not just the management CPU
+pool).
+
+#### Installer Changes
+
+The installer will be changed to accept the new `managementCPUIDs`
+configuration input value.
+
+The installer will be changed to generate an extra machine config
+manifest to configure CRI-O so that containers from pods with the
+`io.openshift.management.cores` annotation are run on the
+`mgmt_ctr_cpuset`.
+
+The installer will be changed to create a machine config manifest to
+write the `/etc/kubernetes/management-pinning` configuration file for
+kubelet. The file will have SELinux settings configured so that it is
+only readable by the kubelet.
+
+The installer will be changed to update the
+`status.managementWorkloadPartitioning` field of the Infrastructure
+config resource to `enabled` so the admission hook will know when to
+mutate pods.
+
+### Risks and Mitigations
+
+The first version of this feature must be enabled when a cluster is
+deployed to work correctly. However, some of the enabling
+configurations could be modified by an admin on day 2. We need to
+document clearly that this is not supported. In particular, changing
+the configuration and rebooting may cause normal pods to be isolated
+but will not configure kubelet to also partition static pods.
+
+It is possible to build a cluster with the feature enabled and then
+deploy an operator or other workload that should take advantage of the
+feature without first configuring the namespace properly.  We need to
+document that the configuration only applies to pods created after it
+is set, so if an operator is installed before the namespace setting is
+changed the operator or its operands may need to be re-installed or
+restarted.
+
+The schedule for delivering this feature is very aggressive. We have
+tried to minimize the number of places that need complex changes to
+make it more likely that we can meet the deadline.
+
+Many implementation details proposed here require us to carry patches
+to components that we try to keep in sync with upstream versions. This
+design isolates those patches more fully than the previous approaches
+did, which should make carrying the patches easier.  This design also
+provides an opportunity to upstream the feature in a way that other
+approaches did not, which may mean not having to carry the patches
+indefinitely.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. If we think this feature is eventually going to be useful in
+   regular clusters, do we want the settings in the `bootstrapInPlace`
+   section? Should we add a `managementWorkloadPartitioning` section,
+   or something similar, and say for now that it only applies when the
+   single-node deployment approach is used?
+2. What should the new field in the Infrastructure CRD be?
```

Comment:
3. Should we worry about heterogeneous control plane hardware in multi-node clusters? The install-config schema may need to support specifying the same pool as using different CPUs on different nodes in that case.